### PR TITLE
Beefy equivocation: check all the MMR roots

### PR DESCRIPTION
**Pull request by serban300**
Normally, the BEEFY protocol only accepts a single MMR Root entry in a commitment's payload. But to be extra careful, when validating equivocation reports, let's check all the MMR roots, if there are more.

### CODE REVIEW

File: substrate/primitives/consensus/beefy/src/payload.rs [hunk 1]
```diff
@@ -56,13 +56,26 @@ impl Payload {
...
 	/// Returns a decoded payload value under given `id`.
 	///
 	/// In case the value is not there, or it cannot be decoded `None` is returned.
 	pub fn get_decoded<T: Decode>(&self, id: &BeefyPayloadId) -> Option<T> {
 		self.get_raw(id).and_then(|raw| T::decode(&mut &raw[..]).ok())
 	}
 
+	/// Returns all decoded payload values under given `id`.
+	pub fn get_all_decoded<T: Decode>(&self, id: &BeefyPayloadId) -> Vec<Option<T>> {
+		self.get_all_raw(id).iter().map(|raw| T::decode(&mut &raw[..]).ok()).collect()
```
Comment (bkchr) [inline substrate/primitives/consensus/beefy/src/payload.rs:76]:
```suggestion
	pub fn get_all_decoded<T: Decode>(&self, id: &BeefyPayloadId) -> impl Iterator<Item = Option<T>> {
		self.get_all_raw(id).iter().map(|raw| T::decode(&mut &raw[..]).ok())
```

No need to allocate.

File: substrate/primitives/consensus/beefy/src/payload.rs [hunk 2]
```diff
@@ -56,13 +56,26 @@ impl Payload {
 		Some(&self.0[index].1)
 	}
 
+	/// Returns all the raw payloads under given `id`.
+	pub fn get_all_raw(&self, id: &BeefyPayloadId) -> Vec<&Vec<u8>> {
+		self.0
+			.iter()
+			.filter_map(|probe| if &probe.0 != id { return None } else { Some(&probe.1) })
+			.collect()
```
Comment (bkchr) [inline substrate/primitives/consensus/beefy/src/payload.rs:64]:
```suggestion
	pub fn get_all_raw(&self, id: &BeefyPayloadId) -> impl Iterator<Item = &Vec<u8>> {
		self.0
			.iter()
			.filter_map(|probe| if &probe.0 != id { return None } else { Some(&probe.1) })
```

File: substrate/frame/beefy-mmr/src/lib.rs [hunk 3]
```diff
@@ -258,17 +258,31 @@ where
...
-			match commitment.payload.get_decoded::<MerkleRootOf<T>>(&known_payloads::MMR_ROOT_ID) {
-				Some(commitment_root) => commitment_root,
+		let commitment_roots = commitment
+			.payload
+			.get_all_decoded::<MerkleRootOf<T>>(&known_payloads::MMR_ROOT_ID);
+		if commitment_roots.is_empty() {
+			// If the commitment doesn't contain any MMR root, while the proof is valid,
+			// the commitment is invalid
+			return true;
+		}
```
Comment (bkchr) [inline substrate/frame/beefy-mmr/src/lib.rs:268]: If `get_all_decoded` returns a `Iterator` as I have proposed, this need to be replaced with a `bool found_valid = false`.

**Review by bkchr — APPROVED** [review]

File: substrate/primitives/consensus/beefy/src/payload.rs (hunk 1 above)
Comment (serban300) [reply substrate/primitives/consensus/beefy/src/payload.rs]: True. Done.

**Review by acatangiu — APPROVED** [review]

**Comment by acatangiu** [comment]
needs prdoc and is good to go

File: prdoc/pr_5857.prdoc [hunk 4]
```diff
@@ -0,0 +1,19 @@
+# Schema: Polkadot SDK PRDoc Schema (prdoc) v1.0.0
+# See doc at https://raw.githubusercontent.com/paritytech/polkadot-sdk/master/prdoc/schema_user.json
+
+title: Added benchmarks for BEEFY fork voting
```
Comment (acatangiu) [inline prdoc/pr_5857.prdoc:4]:
```suggestion
title: "Beefy equivocation: check all the MMR roots"
```

**Review by acatangiu — CHANGES_REQUESTED** [review]

**Review by acatangiu — APPROVED** [review]

File: prdoc/pr_5857.prdoc (hunk 4 above)
Comment (serban300) [reply prdoc/pr_5857.prdoc]: Thanks ! Missed that. Started from an old prdoc.

**Pull request merged**